### PR TITLE
#1672 Custom Default Version Policy

### LINKED ISSUE #1672: More Flexible VersionPolicy
## New Feature
Configurable `HttpRequestMessage.VersionPolicy`.

## Motivation for New Feature

The following setup:
- an MVC ASP.NET Core as frontend.
- Ocelot gateway
- some microservices

All components above configured to use Kestrel with the following snippet:
```csharp
var builder = WebApplication.CreateBuilder(args);

builder.WebHost.ConfigureKestrel(opts =>
{
    opts.ConfigureEndpointDefaults(lopts =>
    {
        lopts.Protocols = HttpProtocols.Http2;
    });
});
```
All components above should only talk to each other using HTTP/2 and no TLS (plain HTTP).

## User Scenario
With plain Ocelot 18.0.0, I couldn't make them talk to each other with plain HTTP from frontend to the service.
The services will print out error messages like the following:
````
HTTP/2 connection error (PROTOCOL_ERROR): Invalid HTTP/2 connection preface
````

## Solution
Found out that I need to make sure that `HttpRequestMessage` has its `VersionPolicy` to be set `RequestVersionOrHigher`.

## Changes I made

- namespace: `Ocelot.Request.Mapper`
  class: `RequestMapper`
  method: `Task<Response<HttpRequestMessage>> Map(HttpRequest request, DownstreamRoute downstreamRoute)`
  changes:
  - added `VersionPolicy = HttpVersionPolicy.RequestVersionOrHigher,` in `HttpRequestMessage` instantiation.
- namespace: `Ocelot.Requester`
  class: `HttpClientBuilder`
  method: `IHttpClient Create(DownstreamRoute downstreamRoute)`
  changes:
  - added `DefaultRequestVersion = downstreamRoute.DownstreamHttpVersion,`  and `DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher,` in `HttpClient` instantiation.

Then all those components above can talk to each other in HTTP/2.

And I think `DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher` could be `DefaultVersionPolicy downstreamRoute.DownstreamHttpVersionPolicy`.

And I am willing to make a PR if you fine folks think this kind of changes is welcome.

## Specifications

  - Version: 18.0.0
  - Platform: ASP.NET Core 6.0
  - Subsystem: ASP.NET 6.0 Docker image. Should be Debian, yes?

### PR DESCRIPTION
**Pull request by ibnuda**
## Closes #1672 
- #1672 

## New Feature
Configurable `HttpRequestMessage.VersionPolicy`.

## Motivation for New Feature

The following setup:
- an MVC ASP.NET Core as frontend.
- Ocelot gateway
- some microservices

All components above configured to use Kestrel with the following snippet:
```csharp
var builder = WebApplication.CreateBuilder(args);

builder.WebHost.ConfigureKestrel(opts =>
{
    opts.ConfigureEndpointDefaults(lopts =>
    {
        lopts.Protocols = HttpProtocols.Http2;
    });
});
```
All components above should only talk to each other using HTTP/2 and no TLS (plain HTTP).

## User Scenario
With plain Ocelot 18.0.0, I couldn't make them talk to each other with plain HTTP from frontend to the service.
The services will print out error messages like the following:
````
HTTP/2 connection error (PROTOCOL_ERROR): Invalid HTTP/2 connection preface
````

## Solution
Found out that I need to make sure that `HttpRequestMessage` has its `VersionPolicy` to be set `RequestVersionOrHigher`.

## Changes I made

- namespace: `Ocelot.Request.Mapper`
  class: `RequestMapper`
  method: `Task<Response<HttpRequestMessage>> Map(HttpRequest request, DownstreamRoute downstreamRoute)`
  changes:
  - added `VersionPolicy = HttpVersionPolicy.RequestVersionOrHigher,` in `HttpRequestMessage` instantiation.
- namespace: `Ocelot.Requester`
  class: `HttpClientBuilder`
  method: `IHttpClient Create(DownstreamRoute downstreamRoute)`
  changes:
  - added `DefaultRequestVersion = downstreamRoute.DownstreamHttpVersion,`  and `DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher,` in `HttpClient` instantiation.

Then all those components above can talk to each other in HTTP/2.

And I think `DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher` could be `DefaultVersionPolicy downstreamRoute.DownstreamHttpVersionPolicy`.

And I am willing to make a PR if you fine folks think this kind of changes is welcome.

## Specifications

  - Version: 18.0.0
  - Platform: ASP.NET Core 6.0
  - Subsystem: ASP.NET 6.0 Docker image. Should be Debian, yes?

### CODE REVIEW

**Comment by raman-m** [comment]
Hi Ibnu!
Thanks for being proactive!

I'll return to your PR back after merging all PRs with lower numbers... Or will find you a mentor 😉

**Comment by raman-m** [comment]
@ggnaegi  You are the mentor of this PR! Assigned! 😉

File: test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs
Comment (ggnaegi) [inline test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs:21]: Could you maybe test all possible cases?
eg:

```c#

    [Theory]
    [InlineData(VersionPolicyCreator.Downgradable)]
    [InlineData(VersionPolicyCreator.Upgradeable)]
    [InlineData(VersionPolicyCreator.ExactVersion)]
    public void should_create_version_policy_based_on_input(string versionPolicy)
    {
        this.Given(_ => GivenTheInput(versionPolicy))
            .When(_ => WhenICreate())
            .Then(_ => ThenTheResultIs(versionPolicy))
            .BDDfy();
    }

    [Theory]
    [InlineData("")]
    [InlineData(null)]
    [InlineData("string is gibberish")]
    public void should_default_to_request_version_or_lower(string versionPolicy)
    {
        this.Given(_ => GivenTheInput(versionPolicy))
            .When(_ => WhenICreate())
            .Then(_ => ThenTheResultIs(HttpVersionPolicy.RequestVersionOrLower))
            .BDDfy();
    }

private void ThenTheResultIs(string versionPolicy)
    {
        _result.ShouldBe(versionPolicy switch
        {
            VersionPolicyCreator.Upgradeable => HttpVersionPolicy.RequestVersionOrHigher,
            VersionPolicyCreator.ExactVersion => HttpVersionPolicy.RequestVersionExact,
            VersionPolicyCreator.Downgradable => HttpVersionPolicy.RequestVersionOrLower,
            _ => HttpVersionPolicy.RequestVersionOrLower,
        });
    }
```

**Review by ggnaegi — CHANGES_REQUESTED** [review]
I have written some comments, you should add a few more test cases and use string constants for the policies.

**Review by ggnaegi — APPROVED** [review]

**Comment by ibnuda** [comment]
@ggnaegi, sorry for pinging you. i couldn't resolve this conflict. would you please resolve this conflict when you merge this pr? the difference is only whitespace and the intended change.

thanks in advance.

**Comment by raman-m** [comment]
Hello, Ibnu!
Just do resolving conflicts in Visual Studio, because VS has a perfect merge tool. Also, for `src/Ocelot/Configuration/File/FileRoute.cs` conflict I would recommend to use the version from **develop** only, and then re-add your new property.
If no luck, let me know and I'll help you...

**Comment by ibnuda** [comment]
hello raman. thanks, the conflict is resolved.

and might be off-topic. but in ocelot will support net8, yes?

**Comment by raman-m** [comment]
> @ibnuda commented [on Nov 9](#issuecomment-1803094517)
hello raman. thanks, the conflict is resolved.

Thanks!

---
> and might be off-topic. but in ocelot will support net8, yes?

Yes! Soon! FYI #1787

**Comment by raman-m** [comment]
@ggnaegi Is it ready? Or can we add this PR & feature to [Nov'23](../milestone/2) milestone? How confident are you, guys? 😃

**Comment by ggnaegi** [comment]
> @ggnaegi Is it ready? Or can we add this PR & feature to [Nov'23](../milestone/2) milestone? How confident are you, guys? 😃
> 
> I see only unit tests
> 
> * [test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs](/ibnuda/Ocelot/blob/feat/custom-defaultversionpolicy/test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs)

Hello, I think it's safe, we can add it to the november release. Maybe, we could implement an acceptance test...

**Comment by ibnuda** [comment]
hello folks, thanks for the answer.
yes, I'm available to write some acceptance tests. just point me where the rest of the acceptance tests are and the general direction then I will push it up.

**Comment by raman-m** [comment]
> @ggnaegi commented [on Nov 20](#issuecomment-1819971106)
```csharp
namespace Ocelot.AcceptanceTests;
public class DefaultVersionPolicyTests : IDisposable
{ }
```
If we will create all test classes in the root folder then we will make long list of them.
Too many files in root folder! Why not to organize files/namespaces in Acceptance Tests project?
What Ocelot feature will this small feature belong to?

**Comment by ggnaegi** [comment]
> > @ggnaegi commented [on Nov 20](#issuecomment-1819971106)
> 
> ```cs
> namespace Ocelot.AcceptanceTests;
> public class DefaultVersionPolicyTests : IDisposable
> { }
> ```
> 
> If we will create all test classes in the root folder then we will make long list of them. Too many files in root folder! Why not to organize files/namespaces in Acceptance Tests project? What Ocelot feature will this small feature belong to?

Yes, you are right, we should reorganize the tests structure, maybe a new PR?

**Comment by ibnuda** [comment]
hello folks, seems like there's a problem when testing with http2 and https. the test pass just fine in local, but it seems the circle ci setup needs to trust dotnet issued cert.

i made change in the `.circleci/config.yaml` by adding `dotnet dev-certs https` but it seems circleci doesn't pick it up.

what can i do to pass the ci?

thanks in advance.

**Comment by ggnaegi** [comment]
@ibnuda uuupppss... Let me check that

**Comment by ggnaegi** [comment]
Ok, the problem is on ocelot side too... Ok, will have a look later, sorry.

**Comment by ibnuda** [comment]
Thank, ggnaegi. but please don't feel pressured. please do it in your own pace and.

**Comment by raman-m** [comment]
Guys, we have the new commits in `develop`!
Please, rebase feature branch onto **develop**!

**Comment by raman-m** [comment]
> @ibnuda commented [on Nov 22](#issuecomment-1822041131)

What the issue with `.circleci/config.yaml` and certificates? Some builds have failed?
Show the links to the CircleCI logs plz!

**Comment by ibnuda** [comment]
@raman-m here's the link https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot/1401/workflows/fed691a9-ee47-4ecf-8408-58ecdb2802cb/jobs/2652

i will rebase it asap.

**Comment by raman-m** [comment]
@ibnuda  Do you need a help from mentor to rebase feature branch?

**Comment by ggnaegi** [comment]
@ibnuda Ok, so, the tests are now passing, but it wasn't dead easy. https://app.circleci.com/pipelines/circleci/GZQVatYXRpPzhVdjSqemdz/Ub5zkLQTdd3gUF5FDU1Htb/30/workflows/efb521ef-7721-421c-a4e7-9bc1bdacb220/jobs/58
The changes are included mainly in the CI Pipeline image Dockerfile

```Dockerfile
FROM mcr.microsoft.com/dotnet/sdk:7.0-alpine

RUN apk add bash icu-libs krb5-libs libgcc libintl libssl1.1 libstdc++ zlib git openssh-client

# Generate and export the development certificate
RUN dotnet dev-certs https -ep /certs/cert.pem -p '' && \
    chmod 644 /certs/cert.pem

ENV ASPNETCORE_URLS="https://+;http://+" 
ENV ASPNETCORE_HTTPS_PORT=443 
ENV ASPNETCORE_Kestrel__Certificates__Default__Password="" 
ENV ASPNETCORE_Kestrel__Certificates__Default__Path=/certs/cert.pem
```
Still, a simple improvement here:
```c#
  public class VersionPolicies
  {
      public const string RequestVersionExact = nameof(HttpVersionPolicy.RequestVersionExact);
      public const string RequestVersionOrLower = nameof(HttpVersionPolicy.RequestVersionOrLower);
      public const string RequestVersionOrHigher = nameof(HttpVersionPolicy.RequestVersionOrHigher);
  }
```

But you should first rebase and then we will be able to proceed...
From your feature branch you should do ```git rebase develop``` and follow the instructions.

**Comment by raman-m** [comment]
> @ggnaegi commented [on Nov 27](#issuecomment-1828633503)

Are you blocked in development because of required new Docker image for builds?
What is the readiness of this PR to plan to include PR in future milestones?

**Comment by ggnaegi** [comment]
@raman-m 
could you update the image on docker and update the references in the circle ci yaml file on the develop branch?
```dockerfile
FROM mcr.microsoft.com/dotnet/sdk:8.0-alpine

RUN apk add bash icu-libs krb5-libs libgcc libintl libssl1.1 libstdc++ zlib git openssh-client

RUN curl -L --output ./dotnet-install.sh https://dot.net/v1/dotnet-install.sh

RUN chmod u+x ./dotnet-install.sh 

# Install .NET 7 SDK
RUN ./dotnet-install.sh -c 7.0 -i /usr/share/dotnet

# Install .NET 6 SDK
RUN ./dotnet-install.sh -c 6.0 -i /usr/share/dotnet

# Generate and export the development certificate
RUN dotnet dev-certs https -ep /certs/cert.pem -p '' && \
    chmod 644 /certs/cert.pem

ENV ASPNETCORE_URLS="https://+;http://+" 
ENV ASPNETCORE_HTTPS_PORT=443 
ENV ASPNETCORE_Kestrel__Certificates__Default__Password="" 
ENV ASPNETCORE_Kestrel__Certificates__Default__Path=/certs/cert.pem
```

**Comment by raman-m** [comment]
@ggnaegi 
Pay attention that we use new `- image: ocelot2/circleci-build:8.21.0` in **[develop](https://github.com/ThreeMammals/Ocelot/blob/develop/.circleci/config.yml#L7)**! We go away from **mijitt0m** account!
The author or you have to resolve merge conflicts. Then create test image in your account please as I said in prev message!

**Comment by raman-m** [comment]
@ibnuda @ggnaegi  Who will resolve merge conflicts? 😉 
Please merge develop!
Also I'm curious, Is it possible to rebase onto develop? 🤣 see a lot of merge commits...

**Comment by raman-m** [comment]
>@ggnaegi commented [on Nov 30, 2023](https://github.com/ThreeMammals/Ocelot/pull/1673#issuecomment-1834534752)

I will try to find some time these days.... Seems both you and the author are blocked by Docker image upgrading.

**Comment by ggnaegi** [comment]
@raman-m I could create the images, but I don't have the needed permissions on ocelot2.

**Comment by raman-m** [comment]
@ggnaegi  You want too much! :wink:
My answer: https://ocelot-2.slack.com/archives/C0667CH7RPD/p1708412345257399
The quote:
> Personal account doesn't allow to add Collaborators, unfortunately.
> Also, see Docker Pricing: https://www.docker.com/pricing/
> **Conclusion**: To work as a team we must upgrade to Team subscription! And sure we have to convert current [ocelot2](https://hub.docker.com/u/ocelot2) account to Organization one.
> 

**Please Note! Team subscription costs $9 per team member monthly!**

**Comment by raman-m** [comment]
Hi @ggnaegi @ibnuda !
I'm experiencing this Docker build error
image
Any ideas?
Have you tried to build the [docker file](https://github.com/ThreeMammals/Ocelot/pull/1673#issuecomment-1834534752) locally in terminal?
Maybe something changed in last 3 months?

Please run local build and let me know your building results plz 👉 
```sh
docker build --platform linux/amd64 -t ocelot2/circleci-build -f Dockerfile.base .
```

**Comment by raman-m** [comment]
I've downloaded https://dl-cdn.alpinelinux.org/alpine/v3.19/community/x86_64/APKINDEX.tar.gz archive and unpacked it,
Opened file `APKINDEX` in text editor and my search for `libssl1.1` was failed.
image

**Comment by raman-m** [comment]
- [microsoft-dotnet-sdk - Official Image | Docker Hub](https://hub.docker.com/_/microsoft-dotnet-sdk/)

Browsing to this "**Official images for the .NET SDK**" I'd say that `8.0-alpine` tags for "**Linux amd64 Tags**" still exist but they are very fresh being updated in February, a week ago! So, I expect that something has really changed and our old build script doesn't work now.
I will try to use old OS and tags, downgrading version and try Ubuntu base image...

**Comment by ggnaegi** [comment]
> Anyway, this 3rd line of the script must be reviewed
> 
> ```dockerfile
> RUN apk add bash icu-libs krb5-libs libgcc libintl libssl1.1 libstdc++ zlib git openssh-client
> ```
> 
> Also you can get some ideas in [alpine3.19 Dockerfile](https://github.com/dotnet/dotnet-docker/blob/6db5b3fb97961df982b7fffce4acbd94ed629306/src/sdk/8.0/alpine3.19/amd64/Dockerfile) which is the latest version and current default base Docker file of `8.0-alpine` tag.

I might just have written that example without checking it @raman-m

**Comment by raman-m** [comment]
I've [rebased](https://github.com/ThreeMammals/Ocelot/pull/1673#event-11864526385) feature branch and resolve conflicts!
The rest of gaps in diff I'll fix during next code review...
Can't help you guys more... 😫

**Comment by raman-m** [comment]
@ggnaegi  I used [this branch](https://github.com/ThreeMammals/Ocelot/commits/1673-docker/) to play with Docker build
I'll create PR after your thorough testing of Docker script.

**Comment by ggnaegi** [comment]
@raman-m I would like to help here, but it's an insane mess..

**Comment by ggnaegi** [comment]
@raman-m Ok, done, could you please build the new images?

**Comment by raman-m** [comment]
Rebased once again!

**Comment by raman-m** [comment]
### Ready for final code review! ✔️

File: test/Ocelot.AcceptanceTests/PollyQoSTests.cs
Comment (raman-m) [inline test/Ocelot.AcceptanceTests/PollyQoSTests.cs:1]: 🤣 QoS tests in this PR...

File: test/Ocelot.IntegrationTests/AdministrationTests.cs
Comment (raman-m) [inline test/Ocelot.IntegrationTests/AdministrationTests.cs:885]: I recommend to switch off the step!
Not related to the feature! It was a fix to have green build

File: test/Ocelot.UnitTests/Configuration/FileConfigurationSetterTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Configuration/FileConfigurationSetterTests.cs:35]: Re-arrange all params having each param on own line!
Hard to read...

File: test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs:7]: File-scoped namespace for new files always!

File: test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Configuration/VersionPolicyCreatorTests.cs:29]: Don't use BDDfy for unit tests! It is useless!
This is old style, and BDDfy will be removed from unit testing project soon.

Use [AAA pattern](https://www.bing.com/search?q=AAA+pattern+unit+test) ❗

File: test/Ocelot.UnitTests/DownstreamRouteFinder/DownstreamRouteFinderMiddlewareTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/DownstreamRouteFinder/DownstreamRouteFinderMiddlewareTests.cs:42]: `System.Net.Http.HttpVersionPolicy.RequestVersionOrLower` 🤯 
Should be `HttpVersionPolicy.RequestVersionOrLower`

File: test/Ocelot.UnitTests/Polly/PollyQoSResiliencePipelineProviderTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Polly/PollyQoSResiliencePipelineProviderTests.cs:167]: @ggnaegi Gui, did you experiment with QoS tests? 😉  
Just ask me to re-run build in the chat! 🆗 ❔

**Review by raman-m — CHANGES_REQUESTED** [review]
Seems this is my first complete review, but it was hard to review because of fake diffs.

My suggestions below 👇

**Comment by raman-m** [comment]
@ggnaegi @RaynaldM 
So, you both approved the PR... But if you don't mind, let's give a time to the author to fix issues from [my code review](https://github.com/ThreeMammals/Ocelot/pull/1673#pullrequestreview-1983704126)

**Review by raman-m — CHANGES_REQUESTED** [review]
Well... 🤕 
Currently, the commits cannot be merged due to a significant diff between branches. The cause is unclear, but it may stem from multiple incorrect rebases and conflict resolutions. I managed to improve the diff for one file, [src/Ocelot/Configuration/File/FileRoute.cs](https://github.com/ThreeMammals/Ocelot/pull/1673/files#diff-d45b4e2fc43ae120f0ba7b129f6c1ebe980d0ecfdd4147b7a18d41c678c4553c), by applying a binary copy operation, which has proven to be extremely helpful.
In the coming days, I plan to address the remaining file diffs and also tackle the last [code review issues](https://github.com/ThreeMammals/Ocelot/pull/1673#pullrequestreview-1983704126) 😉

File: docs/make.bat
Comment (raman-m) [reply docs/make.bat]: Copied original version from target branch: https://github.com/ThreeMammals/Ocelot/pull/1673/commits/ccc66b2e10ef70667ecf1b0c01ba2c67312bbf88

File: src/Ocelot/Configuration/Creator/IVersionPolicyCreator.cs
Comment (raman-m) [reply src/Ocelot/Configuration/Creator/IVersionPolicyCreator.cs:6]: Fixed in https://github.com/ThreeMammals/Ocelot/pull/1673/commits/63b70938a883389034b0c88e1b099630607aa7a8

**Review by raman-m — CHANGES_REQUESTED** [review]
The PR looks good, however, the documentation update appears to be missing❗ 

### Next steps
- [ ] Update docs: [configuration.rst](https://github.com/ibnuda/Ocelot/blob/feat/custom-defaultversionpolicy/docs/features/configuration.rst) and/or [routing.rst](https://github.com/ibnuda/Ocelot/blob/feat/custom-defaultversionpolicy/docs/features/routing.rst)

File: src/Ocelot/Configuration/Creator/HttpVersionPolicyCreator.cs
Comment (raman-m) [inline src/Ocelot/Configuration/Creator/HttpVersionPolicyCreator.cs:18]: The creator will return the `RequestVersionOrLower` as default value!
But the author, Ibnu, wanted another value, `RequestVersionOrHigher`.

So, here is the quote from description:

>   changes:
>   - added `DefaultRequestVersion = downstreamRoute.DownstreamHttpVersion,`  and `DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher,` in `HttpClient` instantiation.
> 
> Then all those components above can talk to each other in HTTP/2.
> 
> And I think `DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrHigher` could be `DefaultVersionPolicy downstreamRoute.DownstreamHttpVersionPolicy`.
> 

I don't see any code with `DefaultVersionPolicy` name or method. What I see now is that in the switch construction, default section returns:
```C#
_ => HttpVersionPolicy.RequestVersionOrLower,
```
If `string downstreamVersionPolicy` argument of the `Create` is empty, absent property in route opts, or the value is not from the list, then the policy will be `RequestVersionOrLower` by default.
**Is it right to return `RequestVersionOrLower` as default one currently?**
Or did we totally miss this requirement?

**Review by raman-m — APPROVED** [review]
### Ready for delivery ✅ 
- Code review ✔️✔️
- Team approvals ✔️✔️✔️
- Unit testing ✔️ &rarr; `HttpVersionPolicyCreatorTests`
- Acceptance tests ✔️ &rarr; `DefaultVersionPolicyTests`
- Updated docs ✔️ &rarr; [configuration.rst](https://github.com/ThreeMammals/Ocelot/pull/1673/files#diff-9811505cb31b8dd98d7bc299d419b806fb8d89d1524507c2fb7d29e7afb531f8)

#### But my questions: [Q1](https://github.com/ThreeMammals/Ocelot/pull/1673#discussion_r1569202230) and [Q2](https://github.com/ThreeMammals/Ocelot/pull/1673#issuecomment-2061970443) ❓

File: src/Ocelot/Configuration/Creator/HttpVersionPolicyCreator.cs
Comment (ggnaegi) [reply src/Ocelot/Configuration/Creator/HttpVersionPolicyCreator.cs:18]: @raman-m we took the default behavior, as documented by Microsoft here: https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpversionpolicy?view=net-8.0

**Comment by raman-m** [comment]
### 🤣 LoL conflicts! 🤯

**Comment by raman-m** [comment]
@ggnaegi [commented](https://github.com/ThreeMammals/Ocelot/pull/1673#issuecomment-2062037172):
> Yes, it makes sense, you're right

**Renamed!** See commits https://github.com/ThreeMammals/Ocelot/pull/1673/commits/15850610971304e852adb6dffab07816bb32f721 and https://github.com/ThreeMammals/Ocelot/pull/1673/commits/891bb164321450a60e14666135fafba1493f5a4f

File: src/Ocelot/Configuration/Creator/HttpVersionPolicyCreator.cs
Comment (raman-m) [reply src/Ocelot/Configuration/Creator/HttpVersionPolicyCreator.cs:18]: Understood! Not an issue...

**Comment by raman-m** [comment]
@ibnuda Congrats! 🥳

**Comment by raman-m** [comment]
Feature commit in develop &rarr; https://github.com/ThreeMammals/Ocelot/commit/176a4c8b2b84bb9b4bcacb03be112eeb1e8dfdc0

**Pull request merged**
